wsagent/wsagent-local/src/main/java/org/eclipse/che/idle/WorkspaceIdleFilter.java

```diff
@@ -0,0 +1,103 @@
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+
+import org.eclipse.che.WorkspaceIdProvider;
+import org.eclipse.che.api.core.rest.HttpJsonRequestFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Singleton
+public class WorkspaceIdleFilter implements Filter {
+    private static final Logger LOG = LoggerFactory.getLogger(WorkspaceIdleFilter.class);
+
+    private final long timeout;
+    private HttpJsonRequestFactory httpJsonRequestFactory;
+    private String apiEndpoint;
+    private ScheduledFuture<?> future;
+    private ScheduledExecutorService executor;
+    private AtomicBoolean stopped;
+
+    @Inject
+    public WorkspaceIdleFilter(HttpJsonRequestFactory httpJsonRequestFactory, @Named("che.api") String apiEndpoint,
+            @Named("che.machine.ws.agent.inactive.stop.timeout.ms") long timeout) {
+        this.timeout = timeout;
+        this.stopped = new AtomicBoolean();
+        if (timeout > 0) {
+            this.apiEndpoint = apiEndpoint;
+            this.httpJsonRequestFactory = httpJsonRequestFactory;
+            this.executor = Executors.newSingleThreadScheduledExecutor();
+            this.future = executor.schedule(this::sendEvent, timeout, TimeUnit.MILLISECONDS);
+            LOG.info("Idling workspace " + WorkspaceIdProvider.getWorkspaceId() + " scheduled [timeout="
+                    + timeout / 1000 + " seconds]");
+            this.stopped.set(false);
+        } else {
+            this.stopped.set(true);
+        }
+    }
+
+    @Override
+    public void init(FilterConfig filterConfig) throws ServletException {
+    }
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
+            throws IOException, ServletException {
+        if (!stopped.get()) {
+            future.cancel(true);
+            future = executor.schedule(this::sendEvent, timeout, TimeUnit.MILLISECONDS);
+            LOG.info("Idling workspace " + WorkspaceIdProvider.getWorkspaceId() + " scheduled [timeout="
+                    + timeout / 1000 + " seconds]");
+        }
+        chain.doFilter(request, response);
+    }
+
+    private void sendEvent() {
+        try {
+            httpJsonRequestFactory.fromUrl(apiEndpoint + "/workspace/" + WorkspaceIdProvider.getWorkspaceId() + "/runtime")
+                                  .useDeleteMethod()
+                                  .request();
+            LOG.info("Workspace " + WorkspaceIdProvider.getWorkspaceId() + " has been stopped.");
```

Comment:
We don't idle a workspace. We stop it. 
I have used: "Stopping workspace " + WorkspaceIdProvider.getWorkspaceId() + " after " + timeout / 1000 + " seconds of inactivity."